daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/FrontStack.scala

```diff
@@ -8,16 +8,31 @@ import FrontStack.{FQ, FQCons, FQEmpty, FQPrepend}
 
 import scalaz.Equal
 
+import scala.collection.generic.CanBuildFrom
 import scala.collection.mutable
 
-/** A stackk which allows to cons, prepend, and pop in constant time, and generate an ImmArray in linear time.
+/** A stack which allows to cons, prepend, and pop in constant time, and generate an ImmArray in linear time.
   * Very useful when needing to traverse stuff in topological order or similar situations.
   */
 final class FrontStack[+A] private (fq: FQ[A], len: Int) {
 
   /** O(1) */
   def length: Int = len
 
+  /** O(n) */
+  @throws[IndexOutOfBoundsException]
+  def slowApply(ix: Int): A = {
+    if (ix < 0) throw new IndexOutOfBoundsException("ix")
+    val i = iterator
+    @tailrec def lp(ix: Int): A =
+      if (!i.hasNext) throw new IndexOutOfBoundsException("ix")
```

Comment:
Same as above, it's probably worth having a more telling error message.

Reply:
@stefanobaghino-da What do you think of these messages?

```scala
scala> List(3)(-1)
java.lang.IndexOutOfBoundsException: -1
  at scala.collection.LinearSeqOptimized.apply(LinearSeqOptimized.scala:67)
  at scala.collection.LinearSeqOptimized.apply$(LinearSeqOptimized.scala:65)
  at scala.collection.immutable.List.apply(List.scala:89)
  ... 28 elided

scala> List(3)(4)
java.lang.IndexOutOfBoundsException: 4
  at scala.collection.LinearSeqOptimized.apply(LinearSeqOptimized.scala:67)
  at scala.collection.LinearSeqOptimized.apply$(LinearSeqOptimized.scala:65)
  at scala.collection.immutable.List.apply(List.scala:89)
  ... 28 elided
```

Reply:
@S11001001 did you mean: `throw new IndexOutOfBoundsException(ix.toString)`
The index won't be printed out the way it is right now

Reply:
it would say: `java.lang.IndexOutOfBoundsException: ix`

Reply:
@leo-da I'm asking if what List does is good to do here.

Reply:
I understand, but I also think there is a typo. If we just print the invalid index, I think that is perfectly fine. However we don't print the index.

Reply:
@leo-da It's not a typo; I deliberately put "ix" there, and I won't put anything else there until everyone who actually cares about what's there has settled on what should be there.

Reply:
I believe that the index that is being accessed is fine. It's what happens on Scala's `List` and Java's `Array`. At the very least it's familiar. From what I see you fixed it in https://github.com/digital-asset/daml/pull/2053/commits/1699d1a72a141e994b450b4078470a298e34bab3 so LGTM. :slightly_smiling_face: